### PR TITLE
stats: remove storm of refcount bumping during metrics snapshot.

### LINKED ISSUE #43836: stats: excessive ref-count inc/dec causing performance bottleneck
We have anecodotal flame-graphs in one application of Envoy where the destruction of MetricSnapshotImpl on the main thread hot-spots in the ref-count decrement code. Context is that there are a very large numbers of stats, and a decent number of threads.

The hot-spot appears to be due to https://github.com/envoyproxy/envoy/pull/18144 which "temporarily" (in Sept 2021 :)) added these references to the MetricSnapshotImpl when collecting the snapshot as insurance that the stats could not be deleted while iterating over them and passing them to the sink. It promised to remove this when switching to iterators to pass the stats to sinks, but this has not occurred.

And I think making that conversion will be a little difficult (not impossible) because:
 * this is in the public Sink interface which has many derived classes in the Envoy codebase and in private repos; changing that interface to be iterator based rather than assuming vectors are available would be fairly toily. Maybe AI could help.
 * The Store and Scope iterators have to run with a lock held, and that might cause deadlocks if the sink handler needs to do a stats operation that requires a lock. This seems intuitively unlikely; it would probaby be fine. But hard to prove. The Store iterator uses a lock in the Allocator, and the Scope iterator uses a lock in the Store, interestingly. Either way there's a deadlock risk depending on what happens in the sink. This would need to be system-tested for every sink.

Several other possible remedies exist, and we may be able to use more than one of them:
 * make the atomic operations on refcount inc/dec more effiicient; the increment can likely have relaxed ordering but the decrement probably has to have a strong barrier. This needs careful thinking, and I think would be somewhat incremental in benefit. This could be done in combination with the other two options below, and we could also do that for the stat values themselves, which probably can tolerate a more relaxed model.
 * Simply revert #18144 which proclaims itself as unnecessary, because the metrics are collected in the main thread, and the main thread is also where the stats would eventually be reaped if someone drops a scope during the sink process @pradeepcrao WDYT?
 * Instead of holding onto a reference in each stat, just hold onto references to the scopes. The Scope defines the lifetime of the stats held within, and this probably gives us a 100x benefit relative to the current state. I think that at the time #18144 was written, scopes were not held as shared_ptr, so refs to each individual stat were kept. Holding scopes as shared_ptr was added in https://github.com/envoyproxy/envoy/pull/19791 in 2022.

@ggreenway @kyessenov may also have opinions.

### PR DESCRIPTION
**Pull request by jmarantz**
Commit Message:

Changes Envoy’s stats snapshotting/lifetime strategy to avoid per-metric refcount churn during a metrics flush by retaining scope references instead of taking a refcounted handle to every individual stat.

Changes:

Update MetricSnapshotImpl to keep metrics alive by capturing ConstScopeSharedPtrs for all scopes, rather than incrementing refcounts on every stat during snapshot creation.
Restrict Allocator/AllocatorImpl stat-construction APIs (makeCounter/makeGauge/makeTextReadout) to protected, pushing call sites toward Scope APIs and enabling the snapshot optimization assumption (“stats are scope-owned”).
Update related tests/benchmarks to use scope-based stat construction and adjust benchmark ranges.

This change makes the existing bmFlushToSinks benchmark in test/server/server_stats_flush_benchmark_test.cc run 5x faster, especially visible when extended to 10M stats, which is done in this PR, highlighting the problematic metrics flush duration >5s when the interval is 5s. I will note also that the problem with the excessive ref-count churn is likely much worse on ARM though I did not measure that.

Before:
```
Benchmark                                         Time             CPU   Iterations
-----------------------------------------------------------------------------------
bmFlushToSinks/10                             0.002 ms        0.002 ms       299435
bmFlushToSinks/100                            0.011 ms        0.011 ms        61284
bmFlushToSinks/1000                           0.109 ms        0.109 ms         6353
bmFlushToSinks/10000                           1.12 ms         1.12 ms          616
bmFlushToSinks/100000                          27.0 ms         27.0 ms           23
bmFlushToSinks/1000000                          394 ms          394 ms            2
bmFlushToSinks/10000000                        4841 ms         4840 ms            1
```
After:
```
Benchmark                                         Time             CPU   Iterations
-----------------------------------------------------------------------------------
bmFlushToSinks/10                             0.002 ms        0.002 ms       436476
bmFlushToSinks/100                            0.004 ms        0.004 ms       202299
bmFlushToSinks/1000                           0.030 ms        0.030 ms        23043
bmFlushToSinks/10000                          0.376 ms        0.375 ms         1879
bmFlushToSinks/100000                          6.73 ms         6.73 ms          103
bmFlushToSinks/1000000                          100 ms          100 ms            7
bmFlushToSinks/10000000                        1065 ms         1065 ms            1
```

Additional Description:
Risk Level: medium -- changing mechanism of keeping stats alive during metrics sink
Testing: //test/... but, man, Envoy test has gotten super flaky since I last did dev. Things eventually pass when I re-run 5 or so integration tests one at a time, on a clean client. IDK what's up with CI.
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a
Fixes #43836

### CODE REVIEW

File: test/server/admin/prometheus_stats_test.cc [hunk 1]
```diff
@@ -54,32 +54,26 @@ class HistogramWrapper {
...
-  ~PrometheusStatsFormatterTest() override { clearStorage(); }
+  ~PrometheusStatsFormatterTest() override { scope_.reset(); }
 
   void addCounter(const std::string& name, Stats::StatNameTagVector cluster_tags) {
-    Stats::StatNameManagedStorage name_storage(baseName(name, cluster_tags), *symbol_table_);
-    Stats::StatNameManagedStorage tag_extracted_name_storage(name, *symbol_table_);
-    counters_.push_back(alloc_.makeCounter(name_storage.statName(),
-                                           tag_extracted_name_storage.statName(), cluster_tags));
+    counters_.push_back(&scope_->counterFromStatNameWithTags(pool_.add(name), cluster_tags));
   }
```
[the remark this answers is not in the document]
Comment (jmarantz) [reply test/server/admin/prometheus_stats_test.cc:65]: Copilot is wrong, here and below; this compiles fine.

File: test/server/admin/prometheus_stats_test.cc [hunk 2]
```diff
@@ -54,32 +54,26 @@ class HistogramWrapper {
...
 
   void addTextReadout(const std::string& name, const std::string& value,
                       Stats::StatNameTagVector cluster_tags) {
-    Stats::StatNameManagedStorage name_storage(baseName(name, cluster_tags), *symbol_table_);
-    Stats::StatNameManagedStorage tag_extracted_name_storage(name, *symbol_table_);
-    Stats::TextReadoutSharedPtr textReadout = alloc_.makeTextReadout(
-        name_storage.statName(), tag_extracted_name_storage.statName(), cluster_tags);
+    Stats::TextReadoutSharedPtr textReadout =
+        &scope_->textReadoutFromStatNameWithTags(pool_.add(name), cluster_tags);
```
[the remark this answers is not in the document]
Comment (jmarantz) [reply test/server/admin/prometheus_stats_test.cc:76]: copilot is incorrect; this compiles fine.

File: test/server/admin/prometheus_stats_test.cc [hunk 3]
```diff
@@ -54,32 +54,26 @@ class HistogramWrapper {
...
 
   void addGauge(const std::string& name, Stats::StatNameTagVector cluster_tags,
                 Stats::Gauge::ImportMode import_mode = Stats::Gauge::ImportMode::Accumulate) {
-    Stats::StatNameManagedStorage name_storage(baseName(name, cluster_tags), *symbol_table_);
-    Stats::StatNameManagedStorage tag_extracted_name_storage(name, *symbol_table_);
-    gauges_.push_back(alloc_.makeGauge(
-        name_storage.statName(), tag_extracted_name_storage.statName(), cluster_tags, import_mode));
+    gauges_.push_back(
+        &scope_->gaugeFromStatNameWithTags(pool_.add(name), cluster_tags, import_mode));
   }
```
[the remark this answers is not in the document]
Comment (jmarantz) [reply test/server/admin/prometheus_stats_test.cc:71]: copilot is incorrect; this compiles fine.

**Comment by jmarantz** [comment]
@pradeepcrao @ggreenway I want to get feedback on this as a whole, but I think it would be reasonble to ask that I make one refactor before this merges.

I tried to ensure we don't directly generate stats from the Stats::Allocator by making its makeCounter (etc) methods protected. I couldn't make them private because of some integration test infrastructure around notifying counters.

However I think what I should do is:
 * merge Stats::AllocatorImpl and Stats::Allocator into one class, and make the methods non-virtual.
 * change the NotifyingCounter mechnaism in the integration tests to make a delegate for Scope instead of at the Allocator level.

I think what I have here is safe today but someone could work around the restricted access to makeCounter by subclassing; it would be better for that to be private.

But I thought I'd get high level comments on this PR first. I think this may go a long way to improving Envoy performance on ARM, at least with large numbers of stats.

**Comment by agrawroh** [comment]
Amazing! <3

/assign @ggreenway

**Comment by jmarantz** [comment]
I think #43968 should merge first and then this PR will be more obviously correct -- will make makeCounter et al private instead of protected.

**Comment by ggreenway** [comment]
I haven't looked at any code yet, but one thing immediately came to mind: in the stats access logger (it's new), for a matched add/subtract pair (for something like an `active`/concurrency stat, it holds a reference to a gauge separate from any scope, so holding a count on all the scopes doesn't necessarily hold all the metrics in the store (although in nearly all cases it would). I'm not sure the best way to handle that. It's currently being worked on in https://github.com/envoyproxy/envoy/pull/43812.

**Comment by jmarantz** [comment]
How would a metric be added to the store without being held by a scope? Would it call Allocator::makeCounter directly (as this PR attempts to prevent by putting makeCounter in protected section (soon to be private section when https://github.com/envoyproxy/envoy/pull/43968 lands)?

**Comment by jmarantz** [comment]
I haven't read about stats access loggers yet but can't they own a scope for the stats they create/manage?

**Comment by kyessenov** [comment]
Looks nice, thanks for re-working the snapshotting. The stats logger has its own long-living scope, so I think we're fine there. We can now also defer-delete scopes and extend their lifetime to the next flush, which solves the long-standing problem of stats missing when flush period is > add/delete of a scope.

**Comment by ggreenway** [comment]
> How would a metric be added to the store without being held by a scope? Would it call Allocator::makeCounter directly (as this PR attempts to prevent by putting makeCounter in protected section (soon to be private section when #43968 lands)?

It's initially created via a scope, but the scope has eviction enabled, and during an operation the logger holds an additional reference/sharedptr to the metric, so it could be removed from the only scope that holds it, but there could still be a reference/sharedptr.

**Comment by jmarantz** [comment]
I think I need to catch up on what scope eviction is. I hadn't thought about the possiblity someone holding onto a reference to a stat, longer than its scope lives.

If that's the case in the code today for stats access loggers, then I think this PR is not safe :(

Do we need stats access logger's stats to be sinked?

**Comment by kyessenov** [comment]
> I think I need to catch up on what scope eviction is. I hadn't thought about the possiblity someone holding onto a reference to a stat, longer than its scope lives.

This is possible without eviction. I think it's probably not happening in the tree, all stats are allocated off-a-scope which should outlast, but I don't know how to verify it.

> If that's the case in the code today for stats access loggers, then I think this PR is not safe :(

Eviction actually requires not holding a reference to a stat. Gauge is a special case, and they are tricky.

**Comment by jmarantz** [comment]
I'm having some trouble understandinge the life of a stats-access-logger stat and how it interacts with scopes. Also, why are gauges different?

Is there some doc or design or detailed PR description, describing the stats-access-logger feature and how it works?

**Comment by kyessenov** [comment]
> I'm having some trouble understandinge the life of a stats-access-logger stat and how it interacts with scopes. Also, why are gauges different?
> 
> Is there some doc or design or detailed PR description, describing the stats-access-logger feature and how it works?

Gauges are different because we actually want UpDownCounter (which is just like two counters that you take a diff off). So we're effectively overlaying semantics of UpDownCounter over Gauge implementation and there's some friction.

**Comment by jmarantz** [comment]
I'm not really following how the semantics of how Gauge values are adjusted and interpreted impact its lifetime.

I think the centerpiece of this PR is that I was trying to lighten the cost of the lifetime management of indidual stats by holding onto the scopes and assuming that would be sufficient. This premise is defeated by:
```
{ // on main thread
  ScopePtr scope = createScope();
  CounterSharedPtr counter = &scope->counter(...); // bumps ref count
  scope.reset();  // counter still exists and is reachable from stats allocator iteration
}
{ // on worker thread
  counter.reset()  // frees the counter which could race with metrics snapshot.
}
```

I'm mulling over another refactor to try to limit the appearance of CounterSharedPtr et al outside the stats system itself. I think only by creating that boundary (if it's not too late) can I make the leap to saying "if you hold onto a scope, all the stats you found in the Allocator are safe".

Another option I am considering, which might be easier, would be avoid looping over the allocator during metrics collection; I should only loop over scopes and the stats held inside the scopes. I could use a flag bit in the stat to de-dedup. This would have the impact of excluding any stats that are not in scopes from the metrics snapshot. That might be OK.

**Comment by ggreenway** [comment]
All the crazy stuff in the stats access logger is new (hasn't made it into a release yet) and isn't working correctly, and the solutions have all felt convoluted. If we can get a huge efficiency gain for an operation that happens all the time, I'd be tempted to completely re-work how that's done in the stats access logger, or delete the feature from it.

**Comment by kyessenov** [comment]
@ggreenway The problem that Josh highlighted is present regardless of the log stats extension. We need some way to couple scope with stat lifetimes throughout the code base, maybe by returning some struct that shares ScopeSharedPtr with raw pointer to the stat to the workers. I don't think it's safe to merge this change without doing that first, since Josh's example may very well lurk somewhere, and the general deletion order with xDS is highly complex.

**Comment by jmarantz** [comment]
The discussion about the new access logger stats only helped highlight that this is too dependent on the patterns stats are typically used, but leaves time bombs for anyone straying outside that.

Summarizing my plan above, it will be either. 

1. Remove CounterSharedPtr et al from the codebase outside stats impl and tests.
2. Iterate over stats inside scopes while holding a reference to the scopes. This will mean we sink only stats in scopes, which is probably ok.

**Comment by pradeepcrao** [comment]
> 1. Remove CounterSharedPtr et al from the codebase outside stats impl and tests.
> 2. Iterate over stats inside scopes while holding a reference to the scopes. This will mean we sink only stats in scopes, which is probably ok.

This sounds like the right thing to do.

**Comment by ggreenway** [comment]
> 1. Remove CounterSharedPtr et al from the codebase outside stats impl and tests.

I agree this seems like the right plan. However, if this is the plan, now is the time to chime in on https://github.com/envoyproxy/envoy/pull/43812

**Comment by jmarantz** [comment]
I'm leaning toward option 1 (loop over scopes) and not try to hide CounterSharedPtr from the masses. It's pretty ingrained at this point. And it's pretty convenient some of the use-cases to have the stat kept alive after the return.

**Comment by pradeepcrao** [comment]
> I'm leaning toward option 1 (loop over scopes) and not try to hide CounterSharedPtr from the masses. It's pretty ingrained at this point. And it's pretty convenient some of the use-cases to have the stat kept alive after the return.

I think we will need both, not just option 1. Use scopes to ensure the stats are not deleted during the flush operation, but still iterating over stats in the allocator. We improved the flush speed by an order of magnitude by not iterating over scopes and deduping stats using a `StatNameHashSet` in #17909

**Comment by jmarantz** [comment]
Oh thanks for the benchmark data from earlier. It  seems things got a lot slower between the :"with change" numbers on that PR and the "before" on this one, degrading from 100ms for num_stats=1M to 304ms measured this week, that 3x performance loss might have happened in #18144 a few weeks after #17909 was merged.

Let me get a read on how fast we can make the scopes based iteration; I think we can de-dup with a spare flag bit in the stat rather than a hash-set, which might help.

I think there's going to be a huge churn and probably multiple refactoring PRs if we try to remove CounterSharedPtr from the codebase generally.

**Comment by ggreenway** [comment]
A quick grep shows not too many hits, but the dynamic modules ABI might be hard to fix: 

edit: nevermind, I don't think that'll be hard to fix. It's a use of stats_store.iterate() and that's in the signature of the required callback.

```
./server/admin/stats_handler.cc:54:  for (const Stats::CounterSharedPtr& counter : server_.stats().counters()) {
./server/admin/prometheus_stats.cc:883:    const std::vector<Stats::CounterSharedPtr>& counters,
./server/admin/prometheus_stats.cc:950:    const std::vector<Stats::CounterSharedPtr>& counters,
./server/admin/prometheus_stats.cc:963:    const std::vector<Stats::CounterSharedPtr>& counters,
./server/admin/prometheus_stats.cc:981:    const std::vector<Stats::CounterSharedPtr>& counters,
./server/admin/stats_request.cc:141:      renderStat<Stats::CounterSharedPtr>(iter->first, response, variant);
./server/admin/prometheus_stats.h:77:  static uint64_t statsAsPrometheus(const std::vector<Stats::CounterSharedPtr>& counters,
./server/admin/prometheus_stats.h:88:  statsAsPrometheusText(const std::vector<Stats::CounterSharedPtr>& counters,
./server/admin/prometheus_stats.h:97:  statsAsPrometheusProtobuf(const std::vector<Stats::CounterSharedPtr>& counters,
./server/admin/prometheus_stats.h:107:  generateWithOutputFormat(const std::vector<Stats::CounterSharedPtr>& counters,
./server/admin/stats_request.h:18:  using StatOrScopes = absl::variant<ScopeVec, Stats::TextReadoutSharedPtr, Stats::CounterSharedPtr,
./server/server.cc:180:        snapped_counters_.push_back(Stats::CounterSharedPtr(&counter));
./server/server.h:494:  std::vector<Stats::CounterSharedPtr> snapped_counters_;
./extensions/bootstrap/dynamic_modules/abi_impl.cc:99:      [&name_view, &found, value_ptr](const Envoy::Stats::CounterSharedPtr& counter) -> bool {
./common/stats/thread_local_store.cc:85:    removeRejectedStats<CounterSharedPtr>(central_cache->counters_,
./common/stats/thread_local_store.cc:86:                                          [this](const CounterSharedPtr& counter) mutable {
./common/stats/thread_local_store.cc:153:std::vector<CounterSharedPtr> ThreadLocalStoreImpl::counters() const {
./common/stats/thread_local_store.cc:155:  std::vector<CounterSharedPtr> ret;
./common/stats/thread_local_store.cc:157:                 [&ret](Counter& counter) { ret.emplace_back(CounterSharedPtr(&counter)); });
./common/stats/thread_local_store.cc:601:         const StatNameTagVector& tags) -> CounterSharedPtr {
./common/stats/thread_local_store.cc:1105:  StatNameHashMap<CounterSharedPtr> counters_;
./common/stats/isolated_store_impl.cc:28:                       StatNameTagVectorOptConstRef tags) -> CounterSharedPtr {
./common/stats/allocator_impl.h:27:  CounterSharedPtr makeCounter(StatName name, StatName tag_extracted_name,
./common/stats/allocator_impl.h:62:  void markCounterForDeletion(const CounterSharedPtr& counter) override;
./common/stats/allocator_impl.h:105:  // to exist to hold it as (e.g.) a CounterSharedPtr rather than a Counter&
./common/stats/allocator_impl.h:107:  std::vector<CounterSharedPtr> deleted_counters_ ABSL_GUARDED_BY(mutex_);
./common/stats/thread_local_store.h:184:  std::vector<CounterSharedPtr> counters() const override;
./common/stats/thread_local_store.h:282:    StatNameHashMap<CounterSharedPtr> counters_;
./common/stats/thread_local_store.h:602:  // to exist to hold it as (e.g.) a CounterSharedPtr rather than a Counter&
./common/stats/thread_local_store.h:619:  CounterSharedPtr counters_overflow_;
./common/stats/thread_local_store.h:620:  CounterSharedPtr gauges_overflow_;
./common/stats/thread_local_store.h:621:  CounterSharedPtr histograms_overflow_;
./common/stats/allocator_impl.cc:300:CounterSharedPtr AllocatorImpl::makeCounter(StatName name, StatName tag_extracted_name,
./common/stats/allocator_impl.cc:309:  auto counter = CounterSharedPtr(makeCounterInternal(name, tag_extracted_name, stat_name_tags));
./common/stats/allocator_impl.cc:470:void AllocatorImpl::markCounterForDeletion(const CounterSharedPtr& counter) {
./common/stats/isolated_store_impl.h:43:  std::vector<CounterSharedPtr> counters() const override { return counters_.toVector(); }
```

**Comment by pradeepcrao** [comment]
> I think we can de-dup with a spare flag bit in the stat rather than a hash-set, which might help.

Neat!

**Comment by kyessenov** [comment]
You could also use "used" bit since it's cleared anyways during flush. So that could be a marker if you combine the loops.

**Comment by ggreenway** [comment]
Is this ready for review, or is this waiting on further changes?

**Comment by jmarantz** [comment]
No sorry I haven't checked in anything yet to address the race I described yesterday. Working on iot.

I can shift to draft-mode to disambiguate.

**Comment by jmarantz** [comment]
I tried implementing the metrics sink by iterating over scopes and that gives back all the gains from this PR. I tried de-duping in two ways: using an absl::flat_hash_set, and using a bit in the stat (which also needs to be cleared afterward). The flat_hash_set method was slightly slower.

So I think it's option 1. Stay tuned. I think we have to have kind of a soft blockade on filters using shared pointers to stats, but allow this fully in tests.  What I'm thinking is that this will be done by hiding the SharedPtr nicknames in files that are only included in the stats internal infrastructure, with similar definitions exposed for tests.

Note also, there are methods on the Store to return vectors of shared pointers of counters, gauges, etc. I think these also need to be removed from the production API and moved to a test utility. This is not too hard -- just lengthy.

**Comment by jmarantz** [comment]
I guess I should note: iterating over scopes does not provide benefit over the current state on x86. It might have a bigger impact on ARM where atomics are costlier.

**Comment by jmarantz** [comment]
I have drafted #44031 as a checkpointt, which I think (when it is complete) will make this PR safe. 

When that's done, CounterSharedPtr et al will not have any visible usage in production outside the stats system.

We may need to have them in tests, still TBD.

**Comment by jmarantz** [comment]
I put an main-thread assert in the destructor of the helper-class for Counters, Gauges, and TextReadouts and it fired in a few unexpected places, but I think I'm being successful wrangling these into shape.

Basically, ThreadLocalStore's cleanup uses ThreadLocal::runOnAllThreads in a way that it seems to expect that the second lambda passed to it gets run on the main thread. But it ain't necessarily so. It runs on the last thread that ThreadLocal processes.

Re-posting to main thread makes things a lot cleaner.

**Comment by jmarantz** [comment]
I'm still working on solving the places where stat destructors get called from something other than the main or test thread. There's a lot of corner cases in our tests that cause this to happen, and I found some interesting races.

**Comment by jmarantz** [comment]
pinging to keep from getting stale. @etruong42 will be looking at this, or at least unblocking this.
